pkg/controller/statefulset/stateful_set_utils.go

```diff
@@ -215,6 +215,26 @@ func isCreated(pod *v1.Pod) bool {
 	return pod.Status.Phase != ""
 }
 
+// isStatefullyStuck returns true if a pod in a stateful set is stuck
+// due to a previously bad roll-out. We can detect this by checking all of:
+// 1) The pod is in a pending state
+// 2) The pod is at a different revision than the update revision
+// 3) The update strategy is rolling
+// 4) The pod should be updated
+func isStatefullyStuck(set *apps.StatefulSet, pod *v1.Pod, currentRevision, updateRevision *apps.ControllerRevision) bool {
+	return isPending(pod) &&
+		getPodRevision(pod) == currentRevision.Name &&
+		currentRevision.Name != updateRevision.Name &&
+		set.Spec.UpdateStrategy.Type == apps.RollingUpdateStatefulSetStrategyType &&
+		set.Spec.UpdateStrategy.RollingUpdate.Partition != nil &&
+		getOrdinal(pod) >= int(*set.Spec.UpdateStrategy.RollingUpdate.Partition)
+}
```

Comment:
This means that the StatefulSet is in the middle of a rollout. This doesn't necessary mean that the StatefulSet is stuck.

Reply:
@janetkuo Consider this (somewhat contrived) example:
- Deploy a StatefulSet with a valid image tag, e.g. `alpine:latest`, let it roll-out
- Re-deploy that StatefulSet with an invalid image, e.g. `alpine:NotValid`, you'll see a StatefulSet pod enter a pending state, that pod is now stuck
- Change the image tag back to a valid one, e.g. `alpine:latest` - it will never roll-out.
My thought is, if a pod is pending, and we have something newer to roll-out, we can kill the pending pod safely.

Now imagine that we have this same case, but with valid image tags:
We roll out `alpine:3.7` and wait for it to hit...10 containers. Once that's deployed, we start rolling out `alpine:3.8`. During that roll-out, we can see that it's a valid image tag, but for whatever reason the upgrade crashes our system even though the container doesn't crashloop, so we immediately go to revert back to `alpine:3.7`. Any pods that have yet to bring up `alpine:3.8` would be safe to delete as we know we have something newer available.
Do you follow that and does it make sense to you?
Would renaming this function to something more precise... maybe...`isPendingButSafeToDelete()` help?